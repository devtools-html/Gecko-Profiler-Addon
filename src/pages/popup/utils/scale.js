export function makeExponentialScale(rangeStart, rangeEnd) {
  const startExp = Math.log(rangeStart);
  const endExp = Math.log(rangeEnd);
  const fromFractionToValue = frac =>
    Math.exp((1 - frac) * startExp + frac * endExp);
  const fromValueToFraction = value =>
    (Math.log(value) - startExp) / (endExp - startExp);
  const fromFractionToSingleDigitValue = frac => {
    return +fromFractionToValue(frac).toPrecision(1);
  };
  return {
    fromFractionToValue,
    fromValueToFraction,
    fromFractionToSingleDigitValue,
  };
}

function lerp(frac, rangeStart, rangeEnd) {
  return (1 - frac) * rangeStart + frac * rangeEnd;
}

export function clamp(val, min, max) {
  return Math.max(min, Math.min(max, val));
}

export function scaleRangeWithClamping(
  val,
  sourceRangeStart,
  sourceRangeEnd,
  destRangeStart,
  destRangeEnd
) {
  const frac = clamp(
    (val - sourceRangeStart) / (sourceRangeEnd - sourceRangeStart),
    0,
    1
  );
  return lerp(frac, destRangeStart, destRangeEnd);
}
