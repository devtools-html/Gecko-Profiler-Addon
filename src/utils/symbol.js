export function unique(debugName, breakpadId) {
  return [debugName, breakpadId].join(':');
  // return `${debugName}:${breakpadId}`;
}
