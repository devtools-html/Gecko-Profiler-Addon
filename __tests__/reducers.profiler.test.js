import reducer from '../src/pages/background/redux/reducers';
import { unique } from '../src/utils/symbol';

describe('reducers.profiler', () => {
  it('should maintain a consistent default state', () => {
    expect(reducer(undefined, {}).profiler).toMatchSnapshot();
  });

  describe('IS_RUNNING', () => {
    it('should set the running state', () => {
      const state = reducer(undefined, { type: 'IS_RUNNING', data: true });
      expect(state.profiler.isRunning).toBe(true);
      expect(
        reducer(state, { type: 'IS_RUNNING', data: false }).profiler.isRunning
      ).toBe(false);
    });
  });

  describe('PROFILER_CAPTURE', () => {
    it('should do nothing until the status is done', () => {
      const state = reducer(undefined, {});
      expect(
        reducer(state, { type: 'PROFILER_CAPTURE', status: 'start' })
      ).toBe(state);
    });
    it('should attach the profile data to the store', () => {
      const data = { foo: 'bar' };
      const { profiler } = reducer(undefined, {
        type: 'PROFILER_CAPTURE',
        status: 'done',
        data,
      });
      expect(profiler.profile).toEqual(data);
    });
  });

  describe('GET_SYMBOLS', () => {
    it('should do nothing until the status is done', () => {
      const state = reducer(undefined, {});
      expect(reducer(state, { type: 'GET_SYMBOLS', status: 'start' })).toBe(
        state
      );
    });
    it('should generate unique symbol addresses in the symbol table', () => {
      const debugName = 'DEBUG_NAME',
        breakpadId = 'BREAKPAD_ID',
        addresses = 'addresses',
        index = 'index',
        buffer = 'buffer';
      const data = { debugName, breakpadId, addresses, index, buffer };
      const state = reducer(undefined, {});
      state.profiler.symbols = {
        [unique(debugName, breakpadId)]: [addresses, index, buffer],
      };
      expect(
        reducer(state, { type: 'GET_SYMBOLS', status: 'done', data })
      ).toEqual(state);
    });
  });
});
