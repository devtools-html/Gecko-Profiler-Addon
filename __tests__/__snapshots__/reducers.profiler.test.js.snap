// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers.profiler should maintain a consistent default state 1`] = `
Object {
  "isRunning": false,
  "profile": null,
  "symbols": Object {},
}
`;
